hack/verify-staticcheck.sh

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env bash
+
+# Copyright 2014 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${KUBE_ROOT}/hack/lib/init.sh"
+source "${KUBE_ROOT}/hack/lib/util.sh"
+
+kube::golang::verify_go_version
+
+FOCUS="${1:-}"
+
+# See https://staticcheck.io/docs/checks
+CHECKS=(
+  "all"
+  "-S1*"   # Omit code simplifications for now.
+  "-ST1*"  # Mostly stylistic, redundant w/ golint
+)
+export IFS=','; checks="${CHECKS[*]}"; unset IFS
+
+# Ensure that we find the binaries we build before anything else.
+export GOBIN="${KUBE_OUTPUT_BINPATH}"
+PATH="${GOBIN}:${PATH}"
+
+# Install staticcheck from vendor
+echo 'installing staticcheck from vendor'
+go install k8s.io/kubernetes/vendor/honnef.co/go/tools/cmd/staticcheck
+
+
+cd "${KUBE_ROOT}"
+
+# Check that the file is in alphabetical order
+failure_file="${KUBE_ROOT}/hack/.staticcheck_failures"
+kube::util::check-file-in-alphabetical-order "${failure_file}"
+
+export IFS=$'\n'
+# NOTE: when "go list -e ./..." is run within GOPATH, it turns the k8s.io/kubernetes
+# as the prefix, however if we run it outside it returns the full path of the file
+# with a leading underscore. We'll need to support both scenarios for all_packages.
+all_packages=()
+while IFS='' read -r line; do all_packages+=("$line"); done < <(go list -e "${FOCUS:-.}"/... | grep -vE "/(third_party|vendor|staging/src/k8s.io/client-go/pkg|generated|clientset_generated)" | sed -e 's|/staging/src/|/vendor/|' -e 's|^k8s.io/kubernetes/|./|' -e "s|^_\(${KUBE_ROOT}/\)\{0,1\}||")
```

Comment:
```
while IFS='' read -r line; do
    all_packages+=("$line");
done < <( \
    go list -e "${FOCUS:-.}"/... \
    | grep -vE "/(third_party|vendor|staging/src/k8s.io/client-go/pkg|generated|clientset_generated)" \
    | sed -e 's|/staging/src/|/vendor/|' -e 's|^k8s.io/kubernetes/|./|' \
          -e "s|^_\(${KUBE_ROOT}/\)\{0,1\}||" \
)
```

Reply:
We have a script that seems to do something similar: hack/make-rules/helpers/cache_go_dirs.sh - it's really fast (yours takes 0m4.613s, that is 0m0.102s) and it picks up some that yours doesn't (not sure why).

The results are cached in .make/all_go_dirs.mk

Maybe run that script and apply any extra filters you need to it?

Reply:
OK, I tried this and ran into this bug: https://github.com/dominikh/go-tools/issues/565

Not sure why, but can we stick with what I have for now? Alternatively, we could create another cache based on the `go list -e ./...` results, or just filter out the broken package.

Reply:
Managed to get this working by ignoring the specific package that triggered the bug, and also adding a special exception to ignore packages with no buildable files (due to build tags). I think these account for all the differences between the 2 approaches.

Reply:
That bug says "Running staticcheck on files from different directories is not valid,"  but you previously said it's crazy slow otherwise, yeah?

Can we beg and pleade to not make that invalid?